packages/connections/src/utils/validation-errors.ts

```diff
@@ -0,0 +1,222 @@
+      });
+    }
+  }
+  return errors;
+}
+
+function _validateSSHTunnelAndTLS(
+  connectionInfo: ConnectionInfo
+): FormValidationError[] {
+  const errors: FormValidationError[] = [];
+  const connectionString = getConnectionString(connectionInfo);
+  if (connectionInfo.connectionOptions.sshTunnel) {
+    if (
+      (connectionString.searchParams.has('tls') ||
+        connectionString.searchParams.has('ssl')) &&
+      !connectionString.searchParams.has('tlsAllowInvalidHostnames')
+    ) {
+      errors.push({
+        message:
+          'tlsAllowInvalidHostnames is currently required when connecting via SSH tunnel with TLS enabled',
+      });
+    }
+  }
+  return errors;
+}
+function _validateMultipleHostsAndSSHTunnel(
+  connectionInfo: ConnectionInfo
+): FormValidationError[] {
+  const errors: FormValidationError[] = [];
+  const connectionString = getConnectionString(connectionInfo);
+  if (
+    connectionString.hosts.length > 1 &&
+    connectionInfo.connectionOptions.sshTunnel?.host
+  ) {
+    errors.push({
+      message:
+        'Connecting via SSH tunnel to replica sets is not currently supported, specify a single host and select directConnection instead',
+    });
+  }
+  return errors;
+}
+function _validateMongoSrvAndSSHTunnel(
+  connectionInfo: ConnectionInfo
+): FormValidationError[] {
+  const connectionString = getConnectionString(connectionInfo);
+  if (
+    connectionString.isSRV &&
+    connectionInfo.connectionOptions.sshTunnel?.host
+  ) {
+    return [
+      {
+        message:
+          'Connecting via SSH tunnel with mongodb+srv is not currently supported',
+      },
+    ];
+  }
+  return [];
+}
+function _validateLocalhostAndTls(
+  connectionString: ConnectionString
+): FormValidationError[] {
+  return [];
+}
+
+function _validateMongoSrvAndPort(
+  connectionString: ConnectionString
+): FormValidationError[] {
+  const errors: FormValidationError[] = [];
+  if (connectionString.isSRV && connectionString.port) {
+    errors.push({
+      field: 'hostname',
+      message: 'Cannot use mongodb+srv and port',
+    });
+    errors.push({
+      field: 'schema',
+      message: 'Cannot use mongodb+srv and port',
+    });
+  }
+
+  return errors;
+}
+export function _validateUsernameAndPassword(
+  connectionString: ConnectionString
+): FormValidationError[] {
+  const errors: FormValidationError[] = [];
+  const { username, password } = connectionString;
+  if (!connectionString.searchParams.has('authMechanism')) {
+    if (!username && !password) {
+      // this is not necessarily an error.
+      return [];
+    }
+    if (!username) {
+      errors.push({
+        field: 'username',
+        message: 'Username is missing',
+      });
+    }
+
+    if (!password) {
+      errors.push({
+        field: 'password',
+        message: 'Password is missing',
+      });
+    }
+  }
+
+  return errors;
+}
+
+function _isSecure(connectionString: ConnectionString): boolean {
```

Comment:
mongodb+srv implies tls is enabled unless a user intentionally disables it with ssl=false or tls=false in the connection string